patches/server/0752-Fix-entity-type-tags-suggestions-in-selectors.patch

```diff
@@ -0,0 +1,87 @@
+@@ -438,6 +438,11 @@ public class Commands {
+     }
+ 
+     public static <T> RequiredArgumentBuilder<CommandSourceStack, T> argument(String name, ArgumentType<T> type) {
++        // Paper start
++        if (com.destroystokyo.paper.PaperConfig.fixTargetSelectorTagCompletion && type.getClass() == net.minecraft.commands.arguments.EntityArgument.class) {
++            return RequiredArgumentBuilder.<CommandSourceStack, T>argument(name, type).suggests(type::listSuggestions);
++        }
++        // Paper end
+         return RequiredArgumentBuilder.argument(name, type);
+     }
+ 
+diff --git a/src/main/java/net/minecraft/commands/arguments/selector/options/EntitySelectorOptions.java b/src/main/java/net/minecraft/commands/arguments/selector/options/EntitySelectorOptions.java
+index 3a351e36cfd674357363bf17b07b18233d0e39d2..415980ad008e76ca210e2e2b603948de5dca2653 100644
+--- a/src/main/java/net/minecraft/commands/arguments/selector/options/EntitySelectorOptions.java
++++ b/src/main/java/net/minecraft/commands/arguments/selector/options/EntitySelectorOptions.java
+@@ -69,6 +69,19 @@ public class EntitySelectorOptions {
+     public static final DynamicCommandExceptionType ERROR_ENTITY_TYPE_INVALID = new DynamicCommandExceptionType((object) -> {
+         return new TranslatableComponent("argument.entity.options.type.invalid", object);
+     });
++    // Paper start
++    public static final DynamicCommandExceptionType ERROR_ENTITY_TAG_INVALID = new DynamicCommandExceptionType((object) -> {
++        return io.papermc.paper.adventure.PaperAdventure
++            .asVanilla(net.kyori.adventure.text.Component
++                .text("Invalid or unknown entity type tag '" + object + "'")
++                .hoverEvent(net.kyori.adventure.text.event.HoverEvent
++                    .showText(net.kyori.adventure.text.Component
++                        .text("You can disable this error in 'paper.yml'")
++                    )
++                )
++            );
++    });
++    // Paper end
+ 
+     private static void register(String id, EntitySelectorOptions.Modifier handler, Predicate<EntitySelectorParser> condition, Component description) {
+         OPTIONS.put(id, new EntitySelectorOptions.Option(handler, condition, description));
+@@ -316,8 +329,20 @@ public class EntitySelectorOptions {
+ 
+                     if (entitySelectorParser.isTag()) {
+                         ResourceLocation resourceLocation = ResourceLocation.read(entitySelectorParser.getReader());
++                        // Paper start - throw error if invalid entity tag
++                        net.minecraft.tags.Tag<EntityType<?>> tag;
++                        if (com.destroystokyo.paper.PaperConfig.fixTargetSelectorTagCompletion) {
++                            tag = EntityTypeTags.getAllTags().getTag(resourceLocation);
++                        } else {
++                            tag = EntityTypeTags.getAllTags().getTagOrEmpty(resourceLocation);
```

Comment:
I assume this is equivalent to the removed `entity.getServer().getTags().getOrEmpty(Registry.ENTITY_TYPE_REGISTRY).getTagOrEmpty(resourceLocation)`, but even so it might be better to copy the vanilla code verbatim for clarity.